mypyc/irbuild/specialize.py

```diff
@@ -135,6 +135,9 @@ def translate_safe_generator_call(
                     + [builder.accept(arg) for arg in expr.args[1:]]),
                 builder.node_type(expr), expr.line, expr.arg_kinds, expr.arg_names)
         else:
+            if callee.fullname == "builtins.tuple":
+                length = builder.builder.builtin_len(builder.accept(expr.args[0].sequences[0]), expr.line)
```

Comment:
DRAFT here. 
I am confused about:
-  how can I get the Value of the `list` from `GeneratorExpr` since `builtin_len` needs a `Value`
-  Where should I move these code to?

Reply:
You'll need to check that the sequence is something that supports `len()` and indexing. Initially it's okay to require is to be a list. Otherwise you can skip the optimization. It's possible to extend this to tuples at least. You can use `node_type(...)` and `is_list_rprimitive` to check if it's a list.

So this would be optimized:

```py
a = [1, 2, 3]
tuple(f(x) for x in a)  # Works because a is a list
```

This wouldn't work:

```py
a = iter([1, 2, 3])
# Can't optimize -- a is an iterable, and we don't know the length ahead of time
tuple(f(x) for x in a)
```

`builder.accept` is correct way to translate an `Expression` into a `Value`.

Since the `tuple(...)` specialization is going to be somewhat involved, you can create a helper function that you call from here. The helper function can check whether the optimization can be performed and return `None` if we should fall back to the default behavior. You can add the helper function in this same file.

Reply:
Thanks! 
Another question: how can I convert `short_int_rprimitive`(got from `builtin_len`) to `c_pyssize_t_rprimitive`?

Reply:
They are the same size, so you only need to shift right by 1 (`IntOp` with `RIGHT_SHIFT`). You can find examples of `LEFT_SHIFT` -- `RIGHT_SHIFT` is pretty much the same.